stdlib/option.mli

```diff
@@ -0,0 +1,73 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Option values.
+
+    @since NEXT_RELEASE *)
+
+(** {1:options Options} *)
+
+type 'a t = 'a option
+(** The type for option values. Either [Some v] or [None]. *)
+
+val some : 'a -> 'a option
+(** [some v] is [Some v]. *)
+
+val evict : none:'a -> 'a option -> 'a
+(** [evict ~none o] is [none] if [o] is [None] and [v] if [o] is [Some v]. *)
+
+val get : 'a option -> 'a
+(** [get o] is [v] if [o] is [Some v] and @raise Invalid_argument otherwise. *)
+
+val map : ('a -> 'b) -> 'a option -> 'b option
+(** [map f o] is [None] if [o] is [None] and [Some (f v)] is [o] is [Some v]. *)
+
+val flatten : 'a option option -> 'a option
+(** [flatten oo] is [Some v] if [oo] is [Some (Some v)] and [None]
+    otherwise. *)
+
+val iter : ('a -> unit) -> 'a option -> unit
+(** [iter f o] is [f v] if [o] is [Some v] and [()] otherwise. *)
+
+val fold : ('a -> 'b -> 'b) -> 'a option -> 'b -> 'b
+(** [fold f o acc] is [f v acc] if [o] is [Some v] and [acc] otherwise. *)
+
+val hash : ('a -> int) -> 'a option -> int
+(** [hash h o] hashes [o] with the help of [h] on [Some _] values. *)
+
+(** {1:preds Predicates and comparisons} *)
+
+val is_none : 'a option -> bool
+(** [is_none o] is [true] iff [o] is [None]. *)
+
+val is_some : 'a option -> bool
+(** [is_some o] is [true] iff [o] is [Some o]. *)
+
+val equal : ('a -> 'a -> bool) -> 'a option -> 'a option -> bool
+(** [equal eq o0 o1] is [true] iff [o0] and [o1] are both [None] or if
+   they are [Some v0] and [Some v1] and [eq v0 v1] is [true]. *)
+
+val compare : ('a -> 'a -> int) -> 'a option -> 'a option -> int
+(** [compare cmp o0 o1] is a total order on options using [cmp] to compare
+    values wrapped by [Some _]. *)
+
+(** {1:convert Converting} *)
+
+val to_result : none:'e -> 'a option -> ('a, 'e) result
+(** [to_result ~none o] is [Ok v] if [o] is [Some v] and [none] otherwise. *)
+
+val of_result : ('a, _) result -> 'a option
+(** [of_result r] is [Some v] if [r] is [Ok v] and [None] otherwise. *)
+
+val to_seq : 'a option -> 'a Seq.t
+(** [to_seq o] is [o] as a sequence. *)
```

Comment:
Perhaps this is obvious enough for most participants to this discussion, but making it explicit that `Some` (resp. `None`) produces a sequence of length 1 (resp 0) might be useful.